text/0032-minimal-relay.md

```diff
@@ -0,0 +1,196 @@
+# RFC-0032: Minimal Relay
+
+|                 |                                                                               |
+| --------------- | ----------------------------------------------------------------------------- |
+| **Start Date**  | 20 September 2023                                                             |
+| **Description** | Proposal to minimise Relay Chain functionality.                               |
+| **Authors**     | Joe Petrowski, Gavin Wood                                                     |
+
+## Summary
+
+The Relay Chain contains most of the core logic for the Polkadot network. While this was necessary
+prior to the launch of parachains and development of XCM, most of this logic can exist in
+parachains. This is a proposal to migrate several subsystems into system parachains.
+
+## Motivation
+
+Polkadot's scaling approach allows many distinct state machines (known generally as parachains) to
+operate with common guarantees about the validity and security of their state transitions. Polkadot
+provides these common guarantees by executing the state transitions on a strict subset (a backing
+group) of the Relay Chain's validator set.
+
+However, state transitions on the Relay Chain need to be executed by _all_ validators. If any of
+those state transitions can occur on parachains, then the resources of the complement of a single
+backing group could be used to offer more cores. As in, they could be offering more coretime (a.k.a.
+blockspace) to the network.
+
+By minimising state transition logic on the Relay Chain by migrating it into "system chains" -- a
+set of parachains that, with the Relay Chain, make up the Polkadot protocol -- the Polkadot
```

Comment:
We should avoid mixing the general desire to take functionality off of the relay chain with an implementation which gives each function its own chain. We should consider offloading each of these components onto a single chain (staking + balances + identity) for better synchronous composability between them. 

Or to broaden the scope, consider other combinations of system chains which are already deployed.

The underlying goals are to maximize secure blockspace and effectively use blockspace. Anything which requires less than a single full core will have to either use coretime less frequently (poor system UX) or will use the core constantly but inefficiently (resource misallocation).

Reply:
Ideally, we place balances + staking + governance together on one parachain.  We'd still handle era points on the relay chain perhaps, but not dots.  We should figure out if this is possible though, aka how governance controls the relay chain, and how era points and slashing work.

Accumulate being quite flexible and permissionless is one of the two new ideas in corejam.  Accumulate becomes much safer if accumulate cannot perform logic based upon dot balances, but only based upon authorization conditions set by the parachains themselves. 

I'd envision nominations using a chain-forked pattern whatever we do, i.e. the staking chain spins off an nomination chain, to which validators add their opinions.  This requires running parachain blocks longer.

The other of the two new ideas in corejam is to have blocks run a long time by saving their memory state.  Although nice, we think this cannot really help with state migrations or npos, due to how memory accesses work.  We can likely use other techniques to do longer running npos computations on parachains though, like simply allowing them to run longer.

Reply:
Anything that requires non-uniform work package execution times/weights seems impractical. It should be possible to extend computations over multiple blocks, by paging in the relevant data needed for each step of the algorithm at any point. 

Yes, spinning off child chains/tasks for heavy computation such as elections or governance vote-tallying makes a lot of sense, especially since parachain execution is already stateless. The System product benefits a lot from having as many user-exposed functionalities as possible living in the same chain, especially when there is no clear impetus to scale horizontally.

@joepetrowski

Reply:
Once two features are in the same chain, users and applications get very accustomed to their being in the same chain, which makes it difficult to split them into multiple chains. I understand that with today's usage, many of these subsystems could all go into a single system parachain. However, as usage and requirements increase (e.g. several thousand validators), it will hit a bottleneck and need to be split.

IMO it makes more sense to keep these subsystems in separate chains but take advantage of core scheduling to handle resource allocation. We could use one core to process staking, governance, and identity in round-robin fashion (perhaps prioritizing staking at certain times like session changes). You mention degraded system UX but we're still talking 18 second instead of 6 second block times. When a single core no longer meets the execution needs of a subsystem, it's much easier and more agile to allocate a dedicated core to that subsystem than it is to launch a new chain and migrate all the data/logic to it (plus getting parachains to re-address their XCM programs, applications to target new chains, etc.).

RE spinning off child chains for heavy computations, yes this is reasonable but I think it's an optimization. The first priority is to get this work off the Relay Chain. Then we can optimize with things like child chains.

Reply:
> You mention degraded system UX but we're still talking 18 second instead of 6 second block times

With the amount of traffic that an identity chain currently handles, it'd take weeks to make a full block. Occupying a full core even 1/3 of the time is inefficient, UX challenges for users coordinating across many chains notwithstanding.

I see the case for being inefficient with coretime in the face of a coretime surplus, though I expect that the pressures under load would point back to consolidation and the practical experience for end-users of navigating many chains tips it in favor of consolidation now for me.

Full support of moving these pallets off of the relay chain.